src/lavinmq/exchange/exchange.cr

```diff
@@ -167,18 +144,83 @@ module LavinMQ
     end
 
     abstract def type : String
-    abstract def bind(destination : Queue, routing_key : String, headers : AMQP::Table?)
-    abstract def unbind(destination : Queue, routing_key : String, headers : AMQP::Table?)
-    abstract def bind(destination : Exchange, routing_key : String, headers : AMQP::Table?)
-    abstract def unbind(destination : Exchange, routing_key : String, headers : AMQP::Table?)
-    abstract def do_queue_matches(routing_key : String, headers : AMQP::Table?, & : Queue -> _)
-    abstract def do_exchange_matches(routing_key : String, headers : AMQP::Table?, & : Exchange -> _)
-
-    def queue_matches(routing_key : String, headers = nil, &blk : Queue -> _)
+    abstract def bind(destination : Destination, routing_key : String, headers : AMQP::Table?)
+    abstract def unbind(destination : Destination, routing_key : String, headers : AMQP::Table?)
+    abstract def bindings_details : Iterator(BindingDetails)
+
+    def publish(msg : Message, immediate : Bool,
+                queues : Set(Queue) = Set(Queue).new,
+                exchanges : Set(Exchange) = Set(Exchange).new) : Int32
+      @publish_in_count += 1
+      headers = msg.properties.headers
       if should_delay_message?(headers)
-        @delayed_queue.try { |q| yield q }
-      else
-        do_queue_matches(routing_key, headers, &blk)
+        if q = @delayed_queue
+          q.publish(msg)
+          return 1
+        else
+          return 0
+        end
+      end
+      find_queues(msg.routing_key, headers, queues, exchanges)
+      if queues.empty?
+        @unroutable_count += 1
+        return 0
+      end
+      return 0 if immediate && !queues.any? &.immediate_delivery?
+
+      count = 0
+      queues.each do |queue|
+        if queue.publish(msg)
+          @publish_out_count += 1
+          count += 1
+          msg.body_io.seek(-msg.bodysize.to_i64, IO::Seek::Current) # rewind
+        end
+      end
+      count
+    end
+
+    def find_queues(routing_key : String, headers : AMQP::Table?,
+                    queues : Set(Queue),
+                    exchanges : Set(Exchange) = Set(Exchange).new) : Nil
+      return unless exchanges.add? self
+      bindings(routing_key, headers).each do |d|
+        case d
+        when Queue
+          queues.add(d) unless d.closed?
```

Comment:
We also check if the queue is closed when doing the publish

Reply:
it's only cheap if the pointer is in the CPU cache, otherwise it's a pointer deference. `queues` is a `Set` so there's no dup check. That a Queue is in the bindings hash but closed should be a very unlikely condition, like only when the server is shutting down or something.

Reply:
But if the queue is closed, we dont try to insert it so no need for a  check in the `Set` , do you think its faster to try to insert in the `Set` than deref `d` and check if it's closed?

Reply:
it doesn't reduce the number of inserts, as it's a very exceptional case that it is closed but still in bindings. so let's only verify that it is closed or not when we really need it, when doing the publish to the queue.

Reply:
```suggestion
          queues.add(d)
```